File: multicore-c/L2/coloring/src/poi_gen.c
/*
 * POI_GEN
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#include "struct_ext.h"
#include "pcg_ext.h"
#include "mc.h"
#include "cm.h"
#include "rcm.h"
#include "cmrcm.h"
#include "poi_gen.h"

extern int
POI_GEN(FILE *fp21)
{
	int nn, nnp;
	int icN1, icN2, icN3, icN4, icN5, icN6;
	int i, icel, icou, icouG, k;

	nn  = ICELTOT;

	NU = 6;
	NL = 6;

	INL = (int *)calloc(nn, sizeof(int));
	if(INL == NULL) {
		fprintf(stderr, "Error: %s\n", strerror(errno));
		return -1;
	}
	INU = (int *)calloc(nn, sizeof(int));
	if(INU == NULL) {
		fprintf(stderr, "Error: %s\n", strerror(errno));
		return -1;
	}
	IAL = (int **)calloc(nn, sizeof(int *));
	if(IAL == NULL) {
		fprintf(stderr, "Error: %s\n", strerror(errno));
		return -1;
	}
	for(i=0; i<nn; i++) {
		IAL[i] = (int *)calloc(NL, sizeof(int));
		if(IAL[i] == NULL) {
			fprintf(stderr, "Error: %s\n", strerror(errno));
			return -1;
		}
	}
	IAU = (int **)calloc(nn, sizeof(int *));
	if(IAU == NULL) {
		fprintf(stderr, "%s\n", strerror(errno));
		return -1;
	}
	for(i=0; i<nn; i++) {
		IAU[i] = (int *)calloc(NU, sizeof(int));
		if(IAU[i] == NULL) {
			fprintf(stderr, "Error: %s\n", strerror(errno));
			return -1;
		}
	}

/****************
 * CONNECTIVITY *
 ****************/
	for(icel=0; icel<ICELTOT; icel++) {
		icN1 = NEIBcell[icel][0];
		icN2 = NEIBcell[icel][1];
		icN3 = NEIBcell[icel][2];
		icN4 = NEIBcell[icel][3];
		icN5 = NEIBcell[icel][4];
		icN6 = NEIBcell[icel][5];

		icouG = 0;

		if(icN5 != 0) {
			icou = INL[icel] + 1;
			IAL[icel][icou-1] = icN5;
			INL[icel]         = icou;
		}

		if(icN3 != 0) {
			icou = INL[icel] + 1;
			IAL[icel][icou-1] = icN3;
			INL[icel]         = icou;
		}

		if(icN1 != 0) {
			icou = INL[icel] + 1;
			IAL[icel][icou-1] = icN1;
			INL[icel]         = icou;
		}

		if(icN2 != 0) {
			icou = INU[icel] + 1;
			IAU[icel][icou-1] = icN2;
			INU[icel]         = icou;
		}

		if(icN4 != 0) {
			icou = INU[icel] + 1;
			IAU[icel][icou-1] = icN4;
			INU[icel]         = icou;
		}

		if(icN6 != 0) {
			icou = INU[icel] + 1;
			IAU[icel][icou-1] = icN6;
			INU[icel]         = icou;
		}
	}

/************************
 * INITIAL connectivity *
 ************************/
	fprintf(fp21, "### INITIAL connectivity\n");
	for(i=0; i<ICELTOT; i++) {
		fprintf(fp21, "I=%5d     INL(i)=%5d     INU(i)=%5d\n", i+1, INL[i], INU[i]);
		fprintf(fp21, "  IAL:");
		for(k=0; k<INL[i]; k++) {
			fprintf(fp21, "%5d", IAL[i][k]);
			if((k+1)%6 == 0) fprintf(fp21, "\n");
		}
		if(k==0 || k%6) fprintf(fp21, "\n");
		fprintf(fp21, "  IAU:");
		for(k=0; k<INU[i]; k++) {
			fprintf(fp21, "%5d", IAU[i][k]);
			if((k+1)%6 == 0) fprintf(fp21, "\n");
		}
		if(k==0 || k%6) fprintf(fp21, "\n");
	}

/*****************
 * MULTICOLORING *
 *****************/
N111:
	fprintf(stderr, "\n\nYou have%8d elements\n", ICELTOT);
	fprintf(stderr, "How many colors do you need ?\n");
	fprintf(stderr, "  #COLOR must be more than 2 and\n");
	fprintf(stderr, "  #COLOR must not be more than%8d\n", ICELTOT);
	fprintf(stderr, "  if #COLOR= 0 then CM  ordering\n");
	fprintf(stderr, "  if #COLOR=-1 then RCM ordering\n");
	fprintf(stderr, "  if #COLOR<-1 then CMRCM ordering\n");
	fprintf(stderr, "=>\n");
	fscanf(stdin, "%d", &NCOLORtot);
	if(NCOLORtot == 1 && NCOLORtot > ICELTOT) goto N111;

	OLDtoNEW = (int *)calloc(ICELTOT, sizeof(int));
	if(OLDtoNEW == NULL) {
		fprintf(stderr, "Error: %s\n", strerror(errno));
		return -1;
	}
	NEWtoOLD = (int *)calloc(ICELTOT, sizeof(int));
	if(NEWtoOLD == NULL) {
		fprintf(stderr, "Error: %s\n", strerror(errno));
		return -1;
	}
	COLORindex = (int *)calloc(ICELTOT+1, sizeof(int));
	if(COLORindex == NULL) {
		fprintf(stderr, "Error: %s\n", strerror(errno));
		return -1;
	}

	if(NCOLORtot > 0) {
		MC(ICELTOT, NL, NU, INL, IAL, INU, IAU,
				&NCOLORtot, COLORindex, NEWtoOLD, OLDtoNEW);
	} else if(NCOLORtot == 0) {
		CM(ICELTOT, NL, NU, INL, IAL, INU, IAU,
				&NCOLORtot, COLORindex, NEWtoOLD, OLDtoNEW);
	} else if(NCOLORtot ==-1) {
		RCM(ICELTOT, NL, NU, INL, IAL, INU, IAU,
				&NCOLORtot, COLORindex, NEWtoOLD, OLDtoNEW);
	} else if(NCOLORtot <-1) {
		CMRCM(ICELTOT, NL, NU, INL, IAL, INU, IAU,
				&NCOLORtot, COLORindex, NEWtoOLD, OLDtoNEW);
	}

	fprintf(stderr, "\n# TOTAL COLOR number%8d\n", NCOLORtot);

	return 0;
}

File: multicore-f/L2/coloring/src/poi_gen.f
!C
!C***
!C*** POI_GEN
!C***
!C
!C    generate COEF. MATRIX for POISSON equations
!C    
      subroutine POI_GEN

      use STRUCT
      use PCG

      implicit REAL*8 (A-H,O-Z)

!C
!C-- INIT.
      nn = ICELTOT

      NU= 6
      NL= 6

      allocate (INL(nn), INU(nn), IAL(NL,nn), IAU(NU,nn))

      INL= 0
      INU= 0
      IAL= 0
      IAU= 0

!C
!C +--------------+
!C | CONNECTIVITY |
!C +--------------+
!C===
      do icel= 1, ICELTOT
        icN1= NEIBcell(icel,1)
        icN2= NEIBcell(icel,2)
        icN3= NEIBcell(icel,3)
        icN4= NEIBcell(icel,4)
        icN5= NEIBcell(icel,5)
        icN6= NEIBcell(icel,6)

        icouG= 0

        if (icN5.ne.0) then
          icou= INL(icel) + 1
          IAL(icou,icel)= icN5
          INL(     icel)= icou
        endif

        if (icN3.ne.0) then
          icou= INL(icel) + 1
          IAL(icou,icel)= icN3
          INL(     icel)= icou
        endif

        if (icN1.ne.0) then
          icou= INL(icel) + 1
          IAL(icou,icel)= icN1
          INL(     icel)= icou
        endif

        if (icN2.ne.0) then 
          icou= INU(icel) + 1
          IAU(icou,icel)= icN2
          INU(     icel)= icou
        endif

        if (icN4.ne.0) then
          icou= INU(icel) + 1
          IAU(icou,icel)= icN4
          INU(     icel)= icou
        endif

        if (icN6.ne.0) then
          icou= INU(icel) + 1
          IAU(icou,icel)= icN6
          INU(     icel)= icou
        endif

      enddo
!C===

!C
!C +----------------------+
!C | INITIAL connectivity |
!C +----------------------+
!C===
      write (21,'(a)') '### INITIAL connectivity'
      do i= 1, ICELTOT
        write (21,'(a,i5,a,i5,a,i5)') 'I=', i, '     INL(i)=', INL(i),  &
     &                                         '     INU(i)=', INU(i)
        write (21,'(a, 6i5)') '  IAL:', (IAL(k,i), k= 1, INL(i))
        write (21,'(a, 6i5)') '  IAU:', (IAU(k,i), k= 1, INU(i))
      enddo
!C===

!C
!C +---------------+
!C | MULTICOLORING |
!C +---------------+
!C===
 111  continue
      write (*,'(//a,i8,a)') 'You have', ICELTOT, ' elements.'
      write (*,'(  a     )') 'How many colors do you need ?'
      write (*,'(  a     )') '  #COLOR must be more than 2 and'
      write (*,'(  a,i8  )') '  #COLOR must not be more than', ICELTOT
      write (*,'(  a     )') '   CM if #COLOR .eq. 0'
      write (*,'(  a     )') '  RCM if #COLOR .eq.-1'
      write (*,'(  a     )') 'CMRCM if #COLOR .le.-2'
      write (*,'(  a     )') '=>'
      read  (*,*)             NCOLORtot
      if (NCOLORtot.eq.1.or.NCOLORtot.gt.ICELTOT) goto 111

      allocate (OLDtoNEW(ICELTOT), NEWtoOLD(ICELTOT))
      allocate (COLORindex(0:ICELTOT))

      if (NCOLORtot.gt.0) then
        call MC  (ICELTOT, NL, NU, INL, IAL, INU, IAU,                  &
     &            NCOLORtot, COLORindex, NEWtoOLD, OLDtoNEW)
	write (*,'(//a)') '###  MultiCOloring' 
      endif

      if (NCOLORtot.eq.0) then
        call  CM (ICELTOT, NL, NU, INL, IAL, INU, IAU,                  &
     &            NCOLORtot, COLORindex, NEWtoOLD, OLDtoNEW)
	write (*,'(//a)') '###  CM' 
      endif

      if (NCOLORtot.eq.-1) then
        call RCM (ICELTOT, NL, NU, INL, IAL, INU, IAU,                  &
     &            NCOLORtot, COLORindex, NEWtoOLD, OLDtoNEW)
	write (*,'(//a)') '### RCM' 
      endif

      if (NCOLORtot.lt.-1) then
        call CMRCM (ICELTOT, NL, NU, INL, IAL, INU, IAU,                &
     &            NCOLORtot, COLORindex, NEWtoOLD, OLDtoNEW)
	write (*,'(//a)') '### CM-RCM' 
      endif
!C===

      write (*,'(/a, i8)') '# TOTAL COLOR number', NCOLORtot

      return
      end
